internal/buffer/search.go

```diff
@@ -21,32 +66,23 @@ func (b *Buffer) findDown(r *regexp.Regexp, start, end Loc) ([2]Loc, bool) {
 		start, end = end, start
 	}
 
-	for i := start.Y; i <= end.Y; i++ {
-		l := b.LineBytes(i)
-		charpos := 0
+	rPadded := padRegexp(r)
 
-		if i == start.Y && start.Y == end.Y {
-			nchars := util.CharacterCount(l)
-			start.X = util.Clamp(start.X, 0, nchars)
-			end.X = util.Clamp(end.X, 0, nchars)
-			l = util.SliceStart(l, end.X)
-			l = util.SliceEnd(l, start.X)
-			charpos = start.X
-		} else if i == start.Y {
-			nchars := util.CharacterCount(l)
-			start.X = util.Clamp(start.X, 0, nchars)
-			l = util.SliceEnd(l, start.X)
-			charpos = start.X
-		} else if i == end.Y {
-			nchars := util.CharacterCount(l)
-			end.X = util.Clamp(end.X, 0, nchars)
-			l = util.SliceStart(l, end.X)
-		}
+	for i := start.Y; i <= end.Y; i++ {
+		l, charpos, padMode := findLineParams(b, start, end, i)
 
-		match := r.FindIndex(l)
+		match := rPadded[padMode].FindIndex(l)
 
 		if match != nil {
+			if padMode&padStart != 0 {
+				_, size := utf8.DecodeRune(l[match[0]:])
+				match[0] += size
+			}
 			start := Loc{charpos + util.RunePos(l, match[0]), i}
+			if padMode&padEnd != 0 {
+				_, size := utf8.DecodeLastRune(l[:match[1]])
+				match[1] -= size
+			}
```

Comment:
nit: It seems it would be a bit more readable if we rearrange the lines: the two `if`s together and then `start := ` and `end := ` together.

Reply:
I was mulling over the same point ...